### PR TITLE
fix(rpc): address unwrap hardening

### PR DESCRIPTION
**Pull request by Vedd-Patel**
# Pull Request

## Description
this PR hardens reliability in send-to-address paths by removing panic-prone address parsing and replacing it with explicit validation plus safe error propagation
it also includes clippy-driven cleanup required by strict lint gates

## Type of Change
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Code refactor / performance improvement
- [ ] Documentation update only
- [ ] CI / Docker / Build changes
- [ ] Other (please describe):

## Protocol Version(s) Affected
- [ ] Legacy (ECDSA — `messages.rs`, `contract.rs`, `handlers.rs`)
- [ ] Taproot-Musig2 (`messages2.rs`, `contract2.rs`, `handlers2.rs`)
- [ ] Both
- [x] Neither (infrastructure / tooling only)

> **Note:** When modifying protocol logic, changes must be made to **both** versions unless the
> scope is explicitly version-specific.

## Affected Component(s)
- [ ] **makerd** (background daemon)
- [x] **taker** (CLI client)
- [ ] **maker-cli** (command-line tool)
- [ ] Core protocol / cryptography / library
- [ ] **watch_tower** (contract monitoring & recovery)
- [ ] Docker / deployment scripts
- [x] Tests / test framework
- [ ] Documentation (`docs/`)
- [ ] Other (please specify):

### Code Quality
- [x] I ran `cargo +nightly fmt --all` and committed the result
- [x] I ran `cargo +stable clippy --all-features --lib --bins --tests -- -D warnings` with zero warnings
- [x] I ran `cargo +stable clippy --examples -- -D warnings` with zero warnings
- [x] I ran `RUSTDOCFLAGS="-D warnings" cargo +nightly doc --all-features --document-private-items --no-deps` with zero warnings
- [x] Pre-commit git hook passes (`ln -s ../../git_hooks/pre-commit .git/hooks/pre-commit` if not already set)

### Testing
- [x] All unit tests pass (`cargo test`)
- [ ] Integration tests pass (`cargo test --features integration-test`)
- [ ] Changes were manually tested on **regtest**
- [ ] End-to-end maker ↔ taker swap tested (where applicable)
- [ ] Test-only code is gated behind `#[cfg(feature = "integration-test")]`

### Documentation
- [ ] Relevant files in the `docs/` folder were updated
- [ ] Complex logic is commented

### Security & Privacy (Critical)
- [ ] This change preserves **trustlessness** and **atomic swap guarantees**
- [ ] No regression in **sybil resistance** or **fidelity bond** logic
- [ ] Tor anonymity and P2P message flow were reviewed
- [ ] No new attack vectors or trust assumptions introduced
- [ ] Edge cases and error handling considered
- [ ] ZMQ subscription integrity (raw tx/block feed) is unaffected
- [ ] `integration-test` feature flag is not reachable in production code paths

## How to Test

```bash
# required quality gates
cargo +nightly fmt --all -- --check
cargo +stable clippy --all-features --lib --bins --tests -- -D warnings
cargo +stable clippy --examples -- -D warnings
RUSTDOCFLAGS="-D warnings" cargo +nightly doc --all-features --document-private-items --no-deps

# Unit tests
cargo test

# Optional but recommended for this checklist
cargo test --features integration-test
```

## Summary by CodeRabbit

* **Bug Fixes**
  * Address parsing now enforces the configured Bitcoin network and surfaces clear errors (no panics).
  * Invalid maker addresses are skipped with a warning instead of failing.
  * Transaction submission/signing now validate inputs and propagate errors safely.

* **Refactor**
  * Sorting and iteration logic simplified for deterministic ordering and reduced allocations.

* **Chores**
  * Project ignores macOS Finder metadata (.DS_Store).

* **Tests**
  * Added unit tests for address/network validation.

### CODE REVIEW

File: src/bin/taker.rs [hunk 1]
```diff
@@ -353,30 +353,13 @@ fn main() -> Result<(), TakerError> {
...
-                change_address_type: AddressType::P2TR,
-            };
-
-            let tx = wallet.spend_from_wallet(
-                feerate.unwrap_or(MIN_FEE_RATE),
-                destination,
-                &coins_to_spend,
-            )?;
-
-            let txid = wallet.send_tx(&tx).unwrap();
```
Comment (0xEgao) [inline src/bin/taker.rs:407]: Why did you made this change?

File: src/taker/offers.rs [hunk 2]
```diff
@@ -648,7 +653,7 @@ impl OfferBook {
     /// Makers are included for both Legacy and Taproot requests.
     fn get_bad_makers(&self, protocol: &MakerProtocol) -> Vec<OfferAndAddress> {
         let mut makers = self.makers.clone();
-        makers.sort_by(|a, b| a.address.0.port.len().cmp(&b.address.0.port.len()));
+        makers.sort_by_key(|a| a.address.0.port.len());
```
Comment (0xEgao) [inline src/taker/offers.rs]: why this?

File: src/taker/taproot_swap.rs [hunk 3]
```diff
@@ -115,7 +115,7 @@ impl Taker {
             // Create aggregated MuSig2 pubkey for internal key (allows cooperative key-path spend)
             // Order pubkeys lexicographically to match signing order
             let mut ordered_pubkeys = [my_pubkey, *multisig_pubkey];
-            ordered_pubkeys.sort_by(|a, b| a.inner.serialize().cmp(&b.inner.serialize()));
+            ordered_pubkeys.sort_by_key(|a| a.inner.serialize());
```
Comment (0xEgao) [inline src/taker/taproot_swap.rs:118]: same, reason for this

File: src/bin/taker.rs (hunk 1 above)
Comment (Vedd-Patel) [reply src/bin/taker.rs:407]: this code is performing the full pipelime send mannualy in cli- like this (coin selection-> destination -> construction -> tx building ->broadcasting with a panic-prone `unwrap`) so i replaced that with `wallet.send_to_address(...)` bcz thats safer and cleaner abtraction boundary

File: src/taker/offers.rs (hunk 2 above)
Comment (Vedd-Patel) [reply src/taker/offers.rs]: they're the reason of clippy -D warnings,(mentioned in the clippy warning specifically)
they're functionally same 
old one compares the `a.adress.0.port.len()` and `b.address.().port.len()` manually 
whereas the new one sort by `port.len()` key directly

File: src/taker/taproot_swap.rs (hunk 3 above)
Comment (Vedd-Patel) [reply src/taker/taproot_swap.rs:118]: same as above 
old one compares serialized pubkeys manually
the new one sort by serialized pubkey bytes directly
this new one is lint compliant

**Comment by 0xEgao** [comment]
Thanks for contributing good work.
 Just fix the fmt error and add those changes suggest by copilot.

**Comment by Vedd-Patel** [comment]
@0xEgao , i’ve addressed all the requested changes by copilot and coderabbit, while working on one file(from the changes requested), i also had to update some related files, so i’ve made those changes as well in this PR only, fixed the failing lint(fmt) also!
take a look when you get a moment!

File: src/maker/rpc/server.rs [hunk 4]
```diff
@@ -156,6 +157,13 @@ fn handle_request<M: MakerRpc>(maker: &Arc<M>, socket: &mut TcpStream) -> Result
     Ok(())
 }
 
+fn parse_rpc_destination_address(
+    address: &str,
+    network: Network,
+) -> Result<bitcoin::Address, MakerError> {
+    parse_checked_address(address, network).map_err(MakerError::from)
+}
+
```
Comment (mojoX911) [inline src/maker/rpc/server.rs]: Inline this please. No need a for new function for a one line execution. Map the error at call site as needed.

File: src/maker/rpc/server.rs [hunk 5]
```diff
@@ -201,3 +209,31 @@ pub(crate) fn start_rpc_server<M: MakerRpc>(maker: Arc<M>) -> Result<(), MakerEr
...
+    #[test]
+    fn parse_rpc_destination_address_rejects_wrong_network() {
+        let err =
+            parse_rpc_destination_address("1BoatSLRHtKNngkdXEeobR76b53LETtpyT", Network::Regtest)
+                .expect_err("mainnet address should fail for regtest");
+        assert!(matches!(
+            err,
+            MakerError::Net(NetError::InvalidNetworkAddressDetailed(_))
+        ));
```
Comment (mojoX911) [inline src/maker/rpc/server.rs]: Lets assert for the internal detailed error string also. These two cases should produce different message.

File: src/error.rs [hunk 6]
```diff
@@ -19,6 +19,9 @@ pub enum NetError {
     /// Error caused by an invalid network address.
     InvalidNetworkAddress,
 
+    /// Error caused by an invalid Bitcoin address, including parser details.
+    InvalidNetworkAddressDetailed(String),
+
```
Comment (mojoX911) [inline src/error.rs]: This is not really a NetError. This should be within the WalletError. 

Move it there. Rename it to something shorter like `InvalidAddress(String)`.

File: src/maker/rpc/server.rs (hunk 5 above)
Comment (mojoX911) [reply src/maker/rpc/server.rs]: Also similar tests like these are already added in the `utill.rs`. We don't need them here again. Lets consolidate all address parsing tests in the util module.

File: src/utill.rs [hunk 7]
```diff
@@ -375,21 +374,33 @@ pub struct UTXO {
...
 }
 
+/// Parse a user-provided address and enforce the expected network.
+pub(crate) fn parse_checked_address(address: &str, network: Network) -> Result<Address, NetError> {
+    let unchecked = Address::from_str(address)
+        .map_err(|e| NetError::InvalidNetworkAddressDetailed(e.to_string()))?;
+    unchecked.require_network(network).map_err(|_| {
+        NetError::InvalidNetworkAddressDetailed(format!(
+            "address is not valid for the expected Bitcoin network: expected {:?}",
+            network
```
Comment (mojoX911) [inline src/utill.rs]: We can simplify this and retain the internal error message. 


```suggestion
    unchecked.require_network(network).map_err(|e| {
        NetError::InvalidNetworkAddressDetailed(format!(
            "Address Network Missmatch | Expected : {network} | Details: {e}"
        ))
```

File: src/error.rs (hunk 6 above)
Comment (mojoX911) [reply src/error.rs]: A better choice here is to just wrap the internal `PerseError` instead of turning it into a string. The parse error is already quite expressive. So just displaying that would be enough. 

It will be like `WalletError::InvalidAddress(ParseError)`.

File: src/utill.rs [hunk 8]
```diff
@@ -1125,4 +1132,24 @@ mod tests {
...
+        let addr = parse_checked_address("1BoatSLRHtKNngkdXEeobR76b53LETtpyT", Network::Bitcoin);
+        assert!(addr.is_ok());
+    }
+
+    #[test]
+    fn test_parse_checked_address_rejects_wrong_network() {
+        let err = parse_checked_address("1BoatSLRHtKNngkdXEeobR76b53LETtpyT", Network::Testnet)
+            .expect_err("mainnet address should fail on testnet");
+        assert!(matches!(err, NetError::InvalidNetworkAddressDetailed(_)));
+    }
```
Comment (mojoX911) [inline src/utill.rs:1229]: In these tests try to hit as many parse errors as you can. Then assert the internal errors.

**Review by mojoX911 — CHANGES_REQUESTED** [review]
Looks good. Need some more cleanups.

File: src/maker/rpc/server.rs (hunk 4 above)
Comment (Vedd-Patel) [reply src/maker/rpc/server.rs]: done 👍

File: src/maker/rpc/server.rs (hunk 5 above)
Comment (Vedd-Patel) [reply src/maker/rpc/server.rs]: i've moved these tests!

File: src/utill.rs (hunk 8 above)
Comment (Vedd-Patel) [reply src/utill.rs:1229]: i've tried to cover maximum testss as i can, pls let me know if you've any other suggestions for this!
also a declaration, i've used ai for writing comments for the respective tests(only comments, not anything else)

File: src/maker/rpc/server.rs [hunk 9]
```diff
@@ -93,8 +93,11 @@ fn handle_request<M: MakerRpc>(maker: &Arc<M>, socket: &mut TcpStream) -> Result
             feerate,
         } => {
             let amount = Amount::from_sat(amount);
-            let destination_address =
-                parse_rpc_destination_address(&address, maker.config().network)?;
+
+            // Inline address validation - parse and enforce the configured network.
```
Comment (mojoX911) [inline src/maker/rpc/server.rs]: remove comment. its self explanatory what this call is doing.

File: src/error.rs [hunk 10]
```diff
@@ -19,12 +19,7 @@ pub enum NetError {
     /// Error caused by an invalid network address.
     InvalidNetworkAddress,
 
-    /// Error caused by an invalid Bitcoin address, including parser details.
-    InvalidNetworkAddressDetailed(String),
-
-    /// Error related to CBOR (Concise Binary Object Representation) serialization or deserialization.
-    ///
-    /// This variant wraps a [`serde_cbor::Error`] to provide details about the issue.
+    /// Error related to CBOR serialization or deserialization.
```
Comment (mojoX911) [inline src/error.rs]: please retain the previous full cbor comment here.

**Review by mojoX911 — CHANGES_REQUESTED** [review]
LGTM. Small nits and once conflict is resolved, this is gtg.

File: src/error.rs (hunk 10 above)
Comment (Vedd-Patel) [reply src/error.rs]: Okay, I tried to match other comments above, but I guess it's necessary to explain so will revert the changes

**Comment by mojoX911** [comment]
Hey @Vedd-Patel one more round of rebase pls to resolve conflict.

**Comment by Vedd-Patel** [comment]
@mojoX911 done!

**Review by mojoX911 — APPROVED** [review]
Ack
